Update OpenSSL build configuration based on Mbed (#3451)
openenclave/openenclave: docs/OpenSSLSupport.md

# Open Enclave Support for OpenSSL

The OpenSSL on OE is configured with following options
- no-afalgeng
  - Disable the AF_ALG hardware engine.
- no-autoerrinit
- no-autoload-config
- no-aria
  - Disable ARIA block cipher.
- no-blake2
  - Disable Blake2.
- no-bf
  - Disable Blowfish.
- no-camellia
  - Disable camellia.
- no-cast
  - Disable CAST5 block cipher.
- no-capieng
  - Disable the Microsoft CryptoAPI engine.
- no-cms
  - Disable cryptographic message syntax that handles S/MIME v3.1 mail.
- no-dso
  - Disable the OpenSSL DSO API.
- no-gost
  - Disable the Russian GOST crypto engine that requires dynamic loading.
- no-hw
- no-idea
  - Disable IDEA.
- no-md2
  - Disable MD2.
- no-md4
  - Disable MD4.
- no-mdc2
  - Disable MDC2.
- no-nextprotoneg

>   Disable Next Protocol Negotiation (NPN).

Reviewer: ```suggestion
  - Disable Next Protocol Negotiation (NPN).
```

- no-psk
  - Disable PSK.
- no-poly1305
  - Disable Poly-1305.
- no-rfc3779
  - Disable RFC 3379 (Delegated Path Discovery).
- no-rmd160
  - Disable RIPEMD-160.
- no-seed
  - Disable SEED ciphersuites.
- no-rc4
  - Disable RC4.
- no-shared
- no-sm2

>   - Chinese cryptographic algorithms.
> - no-sm3
>   - Chinese cryptographic algorithms.
> - no-sm4
>   - Chinese cryptographic algorithms

Reviewer: ```suggestion
  - Disable Chinese cryptographic algorithms.
- no-sm3
  - Disable Chinese cryptographic algorithms.
- no-sm4
  - Disable Chinese cryptographic algorithms
```

- no-siphash
  - Disable SipHash.
- no-scrypt
  - Disable scrypt KDF.
- no-srp
  - Disable Secure Remote Password (SRP).
- no-ssl2
- no-ssl3
- no-whirlpool
  - Disable Whirlpool hash.
- no-threads
- no-ui-console

>   - Used by the openssl command-line tool that is not required by OE.

Reviewer: ```suggestion
  - Disable support for the openssl command-line tool that is not required by OE.
```

- no-zlib
  - Disable the ZLIb support.
- --with-rand-seed=none

*Note:* The autoalginit option is required by APIs (e.g., EVP) that retrieve algorithms by name so
can not be disabled.

> OE also explicitly disables the following feature

Reviewer: ```suggestion
OE also explicitly disables the following feature:
```

rand_drbg.h | Partial | OE by default does not depend on the default rand method. Therefore, rand_drbg APIs are supported but have no impact on rand APIs. The drbg test is disabled. Refer to the [unsuppored test list](/tests/openssl/tests.unsupported) for more detail. |
randerr.h | Yes | - |
rc2.h | Yes | - |
rc4.h | No | RC4 is disabled by OE. |
rc5.h | No | RC5 is disabled by default (header is present). |

> ripemd.h | No | RIPEMD-160 is diabled by OE.  |

Reviewer: ```suggestion
ripemd.h | No | RIPEMD-160 is disabled by OE.  |
```

ts.h | Yes | - |
tserr.h | Yes | - |
txt_db.h | Yes | - |

> ui.h | No | Configred with no-ui-console. |

Reviewer: ```suggestion
ui.h | No | Configured with no-ui-console. |
```

uierr.h | Yes | - |

> whrlpool.h | No | Whirlpool is diabled by OE. |

Reviewer: ```suggestion
whrlpool.h | No | Whirlpool is disabled by OE. |
```
